ports/esp32: added WPA-Enterprise (new) (#17234)
micropython/micropython: docs/library/network.WLAN.rst

    argument is passed. Otherwise, query current state if no argument is
    provided. Most other methods require active interface.

.. method:: WLAN.connect(ssid=None, key=None, *, bssid=None, param=value, ...)

    Connect to the specified wireless network, using the specified key.
    If *bssid* is given then the connection will be restricted to the
    access-point with that MAC address (the *ssid* must also be specified
    in this case). 

>     For WPA Enterprise extensions (ESP32 port only) see :ref:`esp32_network_wlan`.

Reviewer: ```suggestion
    Additional keyword arguments are supported on ESP32 port for WPA Enterprise extensions and minimum security level. See :ref:`esp32_network_wlan`.
```

Reviewer: The link still points to the esp32 quickref. Now this part is in the same doc it's probably OK to remove this line altogether.

    argument is passed. Otherwise, query current state if no argument is
    provided. Most other methods require active interface.

.. method:: WLAN.connect(ssid=None, key=None, *, bssid=None, wpa3=False, param=value, ...)

    Connect to the specified wireless network, using the specified key.
    If *bssid* is given then the connection will be restricted to the
    access-point with that MAC address (the *ssid* must also be specified
    in this case). *wpa3* enforces WPA3 authentication and will reject the

>     network if WPA3 is not supported.

Author: This is gone already.

   board definitions enable this in their default configuration. Other builds
   need ``CONFIG_ESP_WIFI_CSI_ENABLED=y`` in the ESP-IDF configuration.

>    Channel State Information (CSI) provides per-packet physical layer channel data

Reviewer 2: This should not be indented, it should stay unchanged.

.. _ESP-IDF Wi-Fi Protocols: https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/wifi.html#wi-fi-protocol-mode
.. _Espressif proprietary "long-range" mode:
.. _Espressif long-range documentation: https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/wifi.html#long-range-lr

> ESP32 Specific Extensions

Reviewer: Thanks for moving this to the bottom of the doc. At the moment this material is both here and in `esp32/quickref` (I saw your question about where is better to place it.)

I'm not sure which is the better place to do it, honestly... `esp32/quickref` makes sense but there is already a bunch of "esp32 only" notes in this file.

Happy with whichever you think is better, but please combine into one place.

   they can be read. Increase ``buffer_size`` in ``csi_enable()`` to reduce
   drops.

.. _esp32_wpa_enterprise_mode:

WPA Enterprise Mode (ESP32 only)
--------------------------------

    The ESP32 port supports WPA2 and WPA3 Enterprise modes (not activated by 

>     default - check the build instructions in ports/esp32/README.md). 

Reviewer 2: This whole section should be dedented.  It should start at column 1.

    ``ntptime.settime()``. 
    
    If an external RTC is not feasible, you can after manually setting the time 
    on first boot and connecting to the Wifi AP retrieve the current time 
    using NTP, set the internal RTC, and then save the NTP time to NVS or a file, 
    e.g. once per day or so. When rebooting, use the last saved time to set the 
    internal RTC, connect, and continue as before. If it's only a short reboot 
    and not years later, this sort of makes the current time reboot safe (not 
    exactly, but good enough).
    
    As a last resort, you may decide to deactivate CONFIG_MBEDTLS_HAVE_TIME_DATE 
    in ``sdkconfig``. This disables all TLS certificate time checks, so make sure 
    you understand the implications.

5 GHz WiFi band_mode (ESP32 only)
---------------------------------

    The WPA Enterprise feature for ESP32 also supports explicitly setting the 
    5 GHz WiFi ``band_mode``. To use this feature, see 
    :ref:`esp32_wpa_enterprise_mode`. 

    Additional parameters:

        * band_mode -- set WiFi band mode (integer, one of
          ``network.WLAN.MODE_2G_ONLY``, ``network.WLAN.MODE_5G_ONLY``,

>           ``network.WLAN.MODE_AUTO`` (default))

Reviewer 2: I don't see any of these constants available to Python.  How is this 5G feature intended to be used?

Reviewer: As I was already planning to submit a follow-up to move this feature out of the Enterprise-only build config, I can clean this up then as well.

    practice is to use a battery buffered (I2C) RTC together with 
    ``ntptime.settime()``. 
    
    If an external RTC is not feasible, you can after manually setting the time 
    on first boot and connecting to the Wifi AP retrieve the current time 
    using NTP, set the internal RTC, and then save the NTP time to NVS or a file, 
    e.g. once per day or so. When rebooting, use the last saved time to set the 
    internal RTC, connect, and continue as before. If it's only a short reboot 
    and not years later, this sort of makes the current time reboot safe (not 
    exactly, but good enough).
    
    As a last resort, you may decide to deactivate CONFIG_MBEDTLS_HAVE_TIME_DATE 
    in ``sdkconfig``. This disables all TLS certificate time checks, so make sure 
    you understand the implications. (In an nutshell: although the net is full 
    of how-to articles how to switch off certificate validity checks, this has 
    to be considered a bad security practice because `it makes each and every 
    connection insecure <https://daniel.haxx.se/blog/2025/02/11/disabling-cert-checks-we-have-not-learned-much/>`_

>     , just like putting your house key under the doormat.)

Reviewer: ```suggestion
    you understand the implications.
```

I appreciate you wanting to warn people, but we generally don't editorialise in the docs. Also, this warning confuses two different matters (disabling certificate time checks vs disabling all certificate checks).

Disabling certificate time checks means that if an attacker has temporary access to a compromised CA that produces a trusted cert then they could force the device to continue trusting that certificate after it should have otherwise expired. There have only been a handful of cases ever where CA trust is known to be somehow compromised (I'm aware of two, in both cases where the CA was signing certs that they shouldn't have been signing). Meaning this is a fairly unlikely attack that requires the attacker to have very significant resources, or for a very major public CA compromise to occur somehow (front page news level).

Disabling all certificate checks, on the other hand, means that anybody with access to set up a spoofed AP or server could compromise the connection. That one is indeed like putting the key under the mat, but it's otherwise not being discussed here.

    first boot after flashing as shown above, then connect to the AP and set the
    time like so, before starting your application proper:
     
        import ntptime
        ntptime.settime() 
    
    Please note that ``ntptime.settime()`` sets the RTC hard, without gradually
    adjusting it. Chances are you end up with the system time jumping. If your
    application depends on a uniform continuous system time, you'll probably 
    want to avoid setting the system time this way during normal operation.

    In this case (or even in general), you can for example retrieve the time 
    in regular intervals (say, on an hourly basis) like this: 
    
        t = ntptime.time() 
    
    and write ``t`` to NVS or to a file. When the MCU is rebooted later, the 
    last known good time can then be set from the file or NVS instead from 
    the build date. This way, your system is continuous during operation, and 
    it kind of survives a reboot (well, close enough).

5 GHz WiFi band_mode (ESP32 only)
---------------------------------

    The WPA Enterprise feature for ESP32 also supports explicitly setting the 
    5 GHz WiFi ``band_mode``. To use this feature, see 

>     :ref:`esp32_wpa_enterprise_mode`. 

Reviewer: Eventually it'd be good to offer this option unconditionally on 5GHz-enabled boards, but I think it's OK to merge it gated behind enterprise mode and then we can enable it by default in a follow-up PR.

Author: Definitely it would be. It is basically independent of WPA Enterprise. ESP-IDF will normally default to the stronger network (this can be 2.4 or 5), and hardly anyone will ever need to explicitly set the band_mode. I kept it there because I see only enterprise users having to explicitly set it. But technically I agree.

   they can be read. Increase ``buffer_size`` in ``csi_enable()`` to reduce
   drops.

WPA Enterprise Mode (ESP32 only)
--------------------------------

> 

Reviewer: Please add a note that this support is not compiled in by default due to code size

Author: done, done.

    The ESP32 port supports WPA2 and WPA3 Enterprise modes, as well as explicit 5 GHz

>     WiFi ``band_mode``.

Reviewer: Could you please make a separate short section about 5GHz support? It's not really part of the Enterprise support.

(I don't mind if you want to keep it in this PR, but the other option would be to put it into a separate PR.)

Author: The ref is cleaned up, and I created a new section in network.WLAN.rst to describe the Wifi band mode. Agreed, a separate PR would make sense, but ripping this apart and creating a new PR requires a bit more planning. On the other hand, it _is_ connected to the WPA Enterprise mode IMHO, because the ESP-IDF usually uses the best mode that STA and AP both support. In most cases this option will not be required at all, only if for some reason in specific (enterprise) networks you want to explicitly use 2.4 or 5 GHz.

        * ca_cert -- the CA certificate (filename, string)

    EAP-TTLS parameters:

        * username -- your network username (string)
        * password -- your network password (string)
        * identity -- anonymous identity (string)
        * ca_cert -- the CA certificate (filename, string)
        * ttls_phase2_method -- TTLS Phase 2 method (integer)

    EAP-TTLS supports the following TTLS Phase 2 methods (1):

        * network.WLAN.EAP_TTLS_PHASE2_EAP (0)
        * network.WLAN.EAP_TTLS_PHASE2_MSCHAPV2 (1)
        * network.WLAN.EAP_TTLS_PHASE2_MSCHAP (2)
        * network.WLAN.EAP_TTLS_PHASE2_PAP (3)
        * network.WLAN.EAP_TTLS_PHASE2_CHAP (4)

    Please note that MSCHAPv2 and CHAP have known security issues and should be avoided.

    EAP-TLS parameters:

        * client_cert -- client certificate filename (string)
        * private_key -- private key filename (string)
        * private_key_password -- private key password (string, optional)
        * disable_time_check -- suppress the validity check for the local client certificate when using EAP-TLS (boolean,
          default False). This option is included for the sake of completeness only. In practice,

>           you want to renew the client certificate before expiry.

Reviewer: Is this a viable workaround for the problem of connecting to WPA Enterprise networks without a battery-backed RTC?

Reviewer: (If so then I think this is a better trade-off than setting the RTC to the build date.)

Author: > Is this a viable workaround for the problem of connecting to WPA Enterprise networks without a battery-backed RTC?

No. EAP-TLS involves using a client certificate for authentication, instead of username and password. This option disables checking the validity of this client certificate in case you can't renew it on an embedded device. Which is common but has to be considered bad security practice ... Viable is not the word that comes to mind ;-) 

The RTC issue deals with the validity of the CA certificate (which needs to be stored on the client) against which the server certificate is checked when authenticating, and affects all EAP methods using server certificates, i.e. all except EAP-PWD.

Reviewer: Understood, thanks.

> Which is common but has to be considered bad security practice ... Viable is not the word that comes to mind ;-)

I agree, but everything we're discussing here is compromising the ideal CA security model by trusting certificates which may have expired. If we could have done it explicitly by disabling a check rather than implicitly by setting the RTC date to a fixed value then this would have been a better trade-off, but alas.

> The RTC issue deals with the validity of the CA certificate (which needs to be stored on the client) against which the server certificate is checked when authenticating,

In a typical certificate chain the CA certificate may be long-lived but it signs the "leaf" certificate (for the website or endpoint) with a shorter signature validity period (90 days in a lot of cases). This is specifically so that old certificates will expire shortly after being rotated, even though the trust in the CA or root certificate remains.

In your experience does WPA Enterprise follow the same pattern for its signature chains? If so then I think the "build date" method will stop working when the endpoint certificate rotates, which may be quite often. (i.e. When reset to the build date MicroPython still considers the CA certificate valid, but it will no longer consider the whole certificate chain valid because the final signature has a "not valid before" date that's after the firmware build date.)

This is based on my understanding of public certificate infra for generic TLS though, maybe WPA Enterprise does things a little differently?

Reviewer: To help illustrate, here's a hypothetical sequence of events:

1. Trusted CA certificate is valid from 1 Jan 2025 to 1 Jan 2030.
2. On 1 July 2026 the CA signs a server certificate valid from 1 July 2026 until 1 October 2026.
3. MicroPython is built on 28 July 2026.
4. MicroPython boot code sets the RTC date to the build date (28 July 2026).
5. MicroPython connects to the server and receives the server's certificate chain. MicroPython trusts the CA certificate and the CA's signature on the server certificate. Therefore it trusts the server.
6. On 1 September 2026 the CA signs a new server certificate valid from 1 September 2026 until 1 December 2026.
7. MicroPython device resets and sets the RTC date to the build date (28 July 2026).
8. MicroPython connects to the server and receives the server's updated certificate chain. MicroPython trusts the CA certificate but no longer trusts the server's supplied certificate, because 28 July 2026 is before 1 September 2026. Therefore it does not trust the server.

Reviewer: Some extra details from looking at how ESP-IDF and Zephyr do this:

* In Zephyr the [EAP crypto functions](https://docs.zephyrproject.org/latest/services/connectivity/networking/api/wifi_crypto.html) are handled by mbedTLS, but [Zephyr defaults to not building mbedTLS with time support](https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/mbedtls/Kconfig.mbedtls#L216). This [prints a warning at build time](https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/mbedtls/CMakeLists.txt#L119) but it looks like the WPA Enterprise samples would not work otherwise.
* In ESP-IDF [mbedTLS expiry checks](https://github.com/espressif/esp-idf/blob/master/components/mbedtls/Kconfig#L140) are also disabled by default. MicroPython enables these. I'm not 100% which checks for WPA Enterprise are done in wpa_supplicant component versus mbedTLS tbh, but I think this is the answer to how the ESP-IDF example code connects to a server without needing the time set.

Reviewer: In light of the above, maybe the "WPA Enterprise" build option for MicroPython should also disable mbedTLS timestamp checks (can be done via an ESP-IDF sdkconfig entry.)

Author: Hmmmm, I am not a subject matter expert on certificates in general, but let me muse about this. Excuse me if this gets lengthy. 

Granted, deactivating CONFIG_MBEDTLS_HAVE_TIME_DATE in sdkconfig would sort of "resolve" the first boot chicken and egg problem of setting the time correctly in order to connect to the Wifi in order to retrieve the correct time using NTP. But it will silently affect all other layers using mbed_tls, and thus open a whole can of worms when it comes to network security. That others chose to deactivate time checks does not tell me they are right, it only tells me that they were not able to solve the actual underlying problem. 

But anyway - if the MicroPython project chooses to follow this route, I'll silently remove my trick in the documentation, and leave it up to you guys to deactivate CONFIG_MBEDTLS_HAVE_TIME_DATE when WPA Enterprise is enabled, and document that and why this is done. It is only that I do not want to take the blame for MicroPython being weakened in general. 

So how do we go from here? 

An option would be to not deactivate CONFIG_MBEDTLS_HAVE_TIME_DATE in general but tell users the implications and leave it up to them, like

"On the very first boot after flashing the device, chances are that your system time is not correct, and the ESP-IDF will complain about a failed authentication. The reason is likely a failing time validity check of the CA certificate that you need to upload to your device (i.e., the system time is outside of the validity period of the certificate). Best practice is to use a battery buffered (I2C) RTC and set its time accurately once (manually) and then using NTP. If this is not feasible, you can after setting the time manually on first boot retrieve the current time using NTP and save it to NVS or a file, e.g. once per day or so, and use the last saved time when rebooting (if it's only a short reboot and not years later) and then again use NTP. As a last resort, you may decide to deactivate CONFIG_MBEDTLS_HAVE_TIME_DATE in sdkconfig but make sure you understand the implications." 

If you want me to do that, I will not be able to keep the Aug 3rd deadline for the milestone, because I'd rather give it a test which I cannot do over the weekend. Give a day or two. 

By the way I found a person calling himself Angus in this thread in the ESP32 forum https://esp32.com/viewtopic.php?t=18817 - that wasn't you by chance, was it? 

In the meantime, I'll fix the other issues.

Author: This being said, using EAP-PWD keeps you clean of this entire mess. In eduroam Wifis this is not an issue because universities and institutions worldwide provide EAP-PWD, and if it's only for Android devices where it is the officially proposed solution. This may be different in company networks, and if they cannot afford a 10 USD PCF8523 breakout board I don't know.

Author: @projectgus  Oh I might even be able to meet the Aug 3rd deadline (as in, end of business day)  due to the time zone difference ;-)

Author: @projectgus OK, so I checked CONFIG_MBEDTLS_HAVE_TIME_DATE today to verify it indeed suppresses certificate time checks with wpa_supplicant when set to 0, and it does. 

Nevertheless I propose to use this as a last resort when nothing else is feasible as far as WPA Enterprise is concerned, because it is a global setting affecting all certificate time checks. So I'd propose to use the "On the very first boot after flashing the device..." paragraph above instead of the build time hack. This leaves the decision to users after all, while a global default setting potentially does not because it may (and will in many cases) go unnoticed. But then, I am not in charge for the MicroPython project as such. 

How should I proceed? 

The other stuff is ready to commit.

Reviewer: > In eduroam Wifis this is not an issue because universities and institutions worldwide provide EAP-PWD, and if it's only for Android devices where it is the officially proposed solution

Thanks, this is very good to know - if there's a commonly deployed certificate-less method then it's a lot easier to keep `CONFIG_MBEDTLS_HAVE_TIME_DATE` enabled. Maybe we should also explicitly recommend EAP-PWD in the docs, if it's available?

Otherwise I think your proposed text for the docs is very good, thanks. Maybe can expand the last sentence to:

*As a last resort, you may decide to deactivate CONFIG_MBEDTLS_HAVE_TIME_DATE in sdkconfig. This disables all TLS certificate time checks, so make sure you understand the implications.*

> By the way I found a person calling himself Angus in this thread in the ESP32 forum https://esp32.com/viewtopic.php?t=18817 - that wasn't you by chance, was it?

Ah yep, that's me! I worked for Espressif between 2016 and 2021. While I was refreshing my understanding I came across [my own commit from 2021](https://github.com/espressif/esp-idf/commit/795e69b5702d68c9a82f0e31f9f176eb4c2b6070) following that forum thread, to add some explanation to the ESP-IDF sdkconfig option.

Author: @projectgus  OK, so I'll finalize this later today so that I hope the deadline will be met.

.. _ESP-IDF Wi-Fi Protocols: https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/wifi.html#wi-fi-protocol-mode
.. _Espressif proprietary "long-range" mode:
.. _Espressif long-range documentation: https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/wifi.html#long-range-lr

> 
> 
> 
> 
> 
> 

Reviewer: ```suggestion
```

Author: done, with all the other cleanups.

Reviewer: Sorry for the nitpick, but these blank lines are still showing for me in the GitHub UI.

Reviewer: Blank lines are still here. I can take these out in a cleanup commit if necessary, though.
